app/models/assignment.rb

```diff
@@ -600,6 +600,76 @@ def summary_json(user)
       numMarked: self.get_num_marked(user.admin? ? nil : user.id) }
   end
 
+  # Generates the summary of the most test results associated with an assignment.
+  def summary_test_results
+    test_groups_query = self.test_groups
+                            .joins(:test_group_results)
+                            .joins(test_group_results: { test_run: :grouping })
+                            .group('test_groups.id', 'groupings.id')
+                            .select('test_groups.id AS test_groups_id',
+                                    'MAX(test_group_results.created_at) AS test_group_results_created_at')
+                            .where('test_runs.submission_id IS NOT NULL')
+                            .to_sql
+
+    TestGroup.joins(test_group_results: :test_results)
+             .joins("INNER JOIN (#{test_groups_query}) sub \
+                     ON test_groups.id = sub.test_groups_id \
+                     AND test_group_results_test_groups.created_at = sub.test_group_results_created_at")
+             .joins(test_group_results: { test_run: { grouping: :group } })
+             .select('test_results.marks_earned as marks_earned',
+                     'test_results.marks_total as marks_total',
+                     'test_results.output',
+                     'test_groups.name',
+                     'test_results.name as test_result_name',
+                     'test_groups_id',
+                     'groups.group_name',
+                     'test_results.status', :extra_info, :error_type)
+  end
+
+  # Generate a JSON summary of the most recent test results associated with an assignment.
+  def summary_test_result_json
+    self.summary_test_results.group_by(&:group_name).transform_values do |grouping|
+      grouping.group_by(&:name)
+    end.to_json
+  end
+
+  # Generate a CSV summary of the most recent test results associated with an assignment.
+  def summary_test_result_csv
+    summary_test_results = self.summary_test_results.as_json
+
+    headers = Set.new
+    groups = Set.new
+    summary_test_results.each do |test_result|
+      groups << test_result['group_name']
+      headers << "#{test_result['group_name']}:#{test_result['test_result_name']}"
+    end
+
+    CSV.generate do |csv|
+      csv << headers
```

Comment:
Hmmm okay, so there's one last data format change I need you to make. The *first column should be the group name* (and the remaining columns should be the actual test results). So for the header row, the first entry should be `nil`, and then followed by the `headers`. For all subsequent rows, the first entry should be the group name, and then the test results for that specific group.